### PR TITLE
Correctly terminate single-job instances

### PR DESCRIPTION
**Pull request by moskyb**
There's a situation that can occur when using the TerminateAfterJob mode of the agent, where when the ASG gets down to its MinSize, after running a job, the agent will attempt to self-terminate using the `aws autoscaling terminate-instance-in-autoscaling-group` command, but becuase that action would cause the ASG to go below its minsize, the call fails. This fails the entire systemd `ExecPostStop` hook, which fails the entire unit, which causes systemd to restart the unit.

This means that in certain circumstances (when the instance attempted to self-terminate but the call failed), single-job instances can retain state from previous jobs, which is the opposite of what we want for these instances.

This PR changes the autoterminate script so that if the instance is in TerminateAfterJob mode, if the call to terminate fails, the instance will call `shutdown` on itself, causing the instance to be removed from the ASG no matter what EC2 has to say about it

### CODE REVIEW

**Review by DrJosh9000 — COMMENTED** [review]
Should we do a similar thing on the Windows side?

**Comment by moskyb** [comment]
> Should we do a similar thing on the Windows side?

ooh, good call, have added now. i'm gonna do some smoke testing before merging, though

File: packer/windows/conf/bin/bk-install-elastic-stack.ps1 [hunk 1]
```diff
@@ -227,6 +228,9 @@ If ((![string]::IsNullOrEmpty($Env:BUILDKITE_ENV_FILE_URL)) -And (Test-Path -Pat
   }
 }
 
+nssm set buildkite-agent AppEnvironmentExtra BUILDKITE_TERMINATE_INSTANCE_AFTER_JOBUILDKITE_TERMINATE_INSTANCE_AFTER_JOB=$Env:BUILDKITE_TERMINATE_INSTANCE_AFTER_JOB
```
Comment (DrJosh9000) [inline packer/windows/conf/bin/bk-install-elastic-stack.ps1:231]: Should it be this?

```suggestion
nssm set buildkite-agent AppEnvironmentExtra BUILDKITE_TERMINATE_INSTANCE_AFTER_JOB=$Env:BUILDKITE_TERMINATE_INSTANCE_AFTER_JOB
```

File: packer/windows/conf/bin/bk-install-elastic-stack.ps1 [hunk 2]
```diff
@@ -246,9 +250,9 @@ cfn-signal `
   --stack "$Env:BUILDKITE_STACK_NAME" `
   --resource "AgentAutoScaleGroup" `
   --exit-code 0 ; if (-not $?) {
-    # This will fail if the stack has already completed, for instance if there is a min size
-    # of 1 and this is the 2nd instance. This is ok, so we just ignore the erro
-    Write-Output "Signal failed"
-  }
+  # This will fail if the stack has already completed, for instance if there is a min size
+  # of 1 and this is the 2nd instance. This is ok, so we just ignore the erro
+  sm   Write-Output "Signal failed"
```
Comment (DrJosh9000) [inline packer/windows/conf/bin/bk-install-elastic-stack.ps1:255]: `sm`?

```suggestion
  Write-Output "Signal failed"
```

File: packer/windows/conf/bin/bk-install-elastic-stack.ps1 (hunk 1 above)
Comment (moskyb) [reply packer/windows/conf/bin/bk-install-elastic-stack.ps1]: no idea what happened here... fixed now though, as with `sm`

File: packer/linux/conf/buildkite-agent/scripts/terminate-instance [hunk 3]
```diff
@@ -12,4 +16,15 @@ region=$(curl -H "X-aws-ec2-metadata-token: $token" --fail --silent --show-error
...
 
-aws autoscaling terminate-instance-in-auto-scaling-group --region "$region" --instance-id "$instance_id" "--should-decrement-desired-capacity"
+if [[ $BUILDKITE_TERMINATE_INSTANCE_AFTER_JOB == "true" ]]; then
+  # If we're the final before the ASG's min size, the call to terminate-instance-in-autoscaling-group will fail, as AWS
+  # won't allow the ASG to go below its min size. In this case, we need to call shutdown instead and force the issue -
+  # the ASG will then spin up a new instance to replace the one we're shutting down, leaving all well in the world.
+  #
+  # We need to do this because if the call to terminate fails, the systemd unit will start up a new buildkite-agent process
+  # on this machine, with state left over from the last agent, which is the opposite of what we want when $BUILDKITE_TERMINATE_INSTANCE_AFTER_JOB is true.
+  terminate "$region" "$instance_id" || shutdown
```
Comment (DrJosh9000) [inline packer/linux/conf/buildkite-agent/scripts/terminate-instance:26]: I think the default time for `shutdown` is 1 minute. `shutdown now`?

File: packer/windows/conf/bin/bk-install-elastic-stack.ps1 [hunk 4]
```diff
@@ -227,6 +228,9 @@ If ((![string]::IsNullOrEmpty($Env:BUILDKITE_ENV_FILE_URL)) -And (Test-Path -Pat
   }
 }
 
+nssm set buildkite-agent AppEnvironmentExtra BUILDKITE_TERMINATE_INSTANCE_AFTER_JOB=$Env:BUILDKITE_TERMINATE_INSTANCE_AFTER_JOB
+If ($lastexitcode -ne 0) { Exit $lastexitcode }
```
Comment (DrJosh9000) [inline packer/windows/conf/bin/bk-install-elastic-stack.ps1:232]: This line repeats the next one - I'm guessing that the existing one was checking the exit code from the `Get-Content` in the loop above. So maybe the lines should be swapped.

**Review by DrJosh9000 — APPROVED** [review]
Some more comments, but I think this is good 🥑

File: packer/windows/conf/bin/bk-install-elastic-stack.ps1
Comment (DrJosh9000) [inline packer/windows/conf/bin/bk-install-elastic-stack.ps1:1]: I commend your effort!

**Comment by hckuo-anyscale** [comment]
when can we merge this?

File: packer/linux/conf/bin/bk-install-elastic-stack.sh [hunk 5]
```diff
@@ -260,6 +260,7 @@ check_docker() {
 check_docker
 
 # start buildkite-agent
+systemctl set-environment "BUILDKITE_TERMINATE_INSTANCE_AFTER_JOB=${BUILDKITE_TERMINATE_INSTANCE_AFTER_JOB}"
```
Comment (triarius) [inline packer/linux/conf/bin/bk-install-elastic-stack.sh:263]: This is going to make it into the job's env yeah? It seems like a useful thing for a job to know, so I'm not opposed to exposing it. But typically, env variables set by the elastic stack are disclosed in an environment hook that sources `/var/lib/buildkite-agent/cfn-env`.

Indeed, that file is the way for configuration set by template parameters to effect the behaviour of various scripts that run on the elastic stack.

You're using a different way of doing that here, which I generally think is better. But it only works for scripts run through systemd. 🤷‍♂️

I'll approve it if it's just addded to `cfn-env` though. That file is generated in https://github.com/buildkite/elastic-ci-stack-for-aws/blob/ee348e4976c9eb4696dc44c3b4f08e59f073059f/packer/linux/conf/bin/bk-install-elastic-stack.sh#L99-L100

File: templates/aws-stack.yml [hunk 6]
```diff
@@ -1092,6 +1092,7 @@ Resources:
           IamInstanceProfile:
             Arn: !GetAtt "IAMInstanceProfile.Arn"
           InstanceType: !Select [ "0", !Split [ ",", !Join [ ",", [ !Ref InstanceTypes, "", "", "" ] ] ] ]
+          InstanceInitiatedShutdownBehavior: "terminate"
```
[the remark this answers is not in the document]
Comment (triarius) [reply templates/aws-stack.yml:1095]: I think if we just make it conditional on the `TerminateInstanceAfterJob` parameter, it's good enough for now.

**Review by triarius — APPROVED** [review]

**Review by triarius — CHANGES_REQUESTED** [review]
I've tested this out with by launching a real stack, and it does not work because the buildkite-agent user does not have permission to shut down the instance.

I'm exploring other options.

**Pull request closed without merging**